File: 剑指offer-2012年书籍版/Java/No13.java
/**
 * 给定单向链表的头指针和一个结点指针， 定义一个函数在O（1）时间删除该节点
 */

public class No13 {

    public static void main(String[] args) {
        MyNode a = new MyNode("A");
        MyNode b = new MyNode("B");
        MyNode c = new MyNode("C");
        MyNode d = new MyNode("D");
        a.setNext(b);
        b.setNext(c);
        c.setNext(d);
        delete(a, d);
        MyNode temp = a;
        while (temp != null) {
            System.out.println(temp.getData());
            temp = temp.next;
        }
    }

    private static void delete(MyNode head, MyNode c) {
        // 如果是尾节点,只能遍历删除
        if (c.next == null) {
            while (head.next != c) {
                head = head.next;
            }
            head.next = null;
        } else if (head == c) {
            head = null;
        } else {
            c.setData(c.getNext().getData());
            c.setNext(c.getNext().getNext());
        }

    }


}

class MyNode {
    MyNode next;
    String data;

    public MyNode(String data) {
        super();
        this.data = data;
    }

    public MyNode getNext() {
        return next;
    }

    public void setNext(MyNode next) {
        this.next = next;
    }

    public String getData() {
        return data;
    }

    public void setData(String data) {
        this.data = data;
    }

}

File: Kotlin/No13.kt
/**
 * 给定单向链表的头指针和一个结点指针， 定义一个函数在O（1）时间删除该节点
 */

class No13 {

    fun main() {
        val node1 = Node("1")
        val node2 = Node("2")
        val node3 = Node("3")
        val node4 = Node("4")
        node1.next = node2
        node2.next = node3
        node3.next = node4

        var head = delete(node1, node3)

        while (head != null) {
            println(head.data)
            head = head.next
        }
    }

    fun delete(head: Node, target: Node): Node? {
        if (target.next == null) {
            // 只能遍历删除
            if (head == target) {
                return null
            } else {
                var temp: Node = head
                while (temp.next != target) {
                    temp = temp.next!!
                }
                temp.next = temp.next?.next
            }
        } else {
            target.data = target.next?.data
            target.next = target.next?.next
        }

        return head
    }
}